service/history/failover/coordinator.go

```diff
@@ -0,0 +1,297 @@
+		case request := <-c.receiveChan:
+			c.handleFailoverMarkers(request)
+		}
+	}
+}
+
+func (c *coordinatorImpl) notifyFailoverMarkerLoop() {
+
+	timer := time.NewTimer(backoff.JitDuration(
+		c.config.FailoverMarkerHeartbeatInterval(),
+		c.config.FailoverMarkerHeartbeatTimerJitterCoefficient(),
+	))
+	defer timer.Stop()
+	requestByMarker := make(map[*replicator.FailoverMarkerAttributes]*receiveRequest)
+
+	for {
+		select {
+		case <-c.shutdownChan:
+			return
+		case notificationReq := <-c.notificationChan:
+			// if there is a shard movement happen, it is fine to have duplicate shard ID in the request
+			// The receiver side will de-dup the shard IDs. See: handleFailoverMarkers
+			for _, marker := range notificationReq.markers {
+				if _, ok := requestByMarker[marker]; !ok {
+					requestByMarker[marker] = &receiveRequest{
+						shardIDs: []int{},
+						marker:   marker,
+						respCh:   notificationReq.respCh,
+					}
+				}
+				req := requestByMarker[marker]
+				req.shardIDs = append(req.shardIDs, notificationReq.shardID)
+			}
+		case <-timer.C:
+			if len(requestByMarker) > 0 {
+				var err error
+				//TODO: Remote calls to send failover markers
+				for marker, request := range requestByMarker {
+					request.respCh <- err
+					close(request.respCh)
+					delete(requestByMarker, marker)
+				}
+			}
+
+			timer.Reset(backoff.JitDuration(
+				c.config.FailoverMarkerHeartbeatInterval(),
+				c.config.FailoverMarkerHeartbeatTimerJitterCoefficient(),
+			))
+		}
+	}
+}
+
+func (c *coordinatorImpl) handleFailoverMarkers(
+	request *receiveRequest,
+) {
+
+	marker := request.marker
+	domainID := marker.GetDomainID()
+
+	if record, ok := c.recorder[domainID]; ok {
+		// if the local failover version is smaller than the new received marker,
+		// it means there is another failover happened and the local one should be invalid.
+		if record.failoverVersion < marker.GetFailoverVersion() {
+			delete(c.recorder, domainID)
+		}
+
+		// if the local failover version is larger than the new received marker,
+		// ignore the incoming marker
+		if record.failoverVersion > marker.GetFailoverVersion() {
+			return
+		}
+	}
+
+	if _, ok := c.recorder[domainID]; !ok {
+		// initialize the failover record
+		c.recorder[marker.GetDomainID()] = &failoverRecord{
+			failoverVersion: marker.GetFailoverVersion(),
+			shards:          make(map[int]struct{}),
+		}
+	}
+
+	record := c.recorder[domainID]
+	record.lastUpdatedTime = c.timeSource.Now()
+	for shardID := range request.shardIDs {
+		record.shards[shardID] = struct{}{}
+	}
+
+	if len(record.shards) == c.config.NumberOfShards {
+		if err := domain.CleanPendingActiveState(
+			c.metadataMgr,
+			domainID,
+			record.failoverVersion,
+			c.retryPolicy,
+		); err != nil {
+			c.logger.Error("Coordinator failed to update domain after receiving all failover markers",
+				tag.WorkflowDomainID(domainID))
+			c.metrics.IncCounter(metrics.DomainFailoverScope, metrics.CadenceFailures)
+			return
```

Comment:
Just want to confirm my understanding here: if `CleanPendingActiveState` failed, all shards will keep notifying the marker and this host will receive new requests from other coordinators so that this operation will be retried?

Reply:
Yes. the pending active state will eventually clean up by the coordinator or after the failove timed out